ztools/x509/extensions.go

```diff
@@ -140,44 +140,153 @@ func (san *SubjectAltName) UnmarshalJSON(b []byte) error {
+	PermittedDNSDomains     []GeneralSubtreeString
+	PermittedEmailDomains   []GeneralSubtreeString
+	PermittedIPAddresses    []GeneralSubtreeIP
+	PermittedDirectoryNames []GeneralSubtreeName
+	PermittedEdiPartyNames  []GeneralSubtreeEdi
+	PermittedRegisteredIDs  []GeneralSubtreeOid
+
+	ExcludedEmailDomains   []GeneralSubtreeString
+	ExcludedDNSDomains     []GeneralSubtreeString
+	ExcludedIPAddresses    []GeneralSubtreeIP
+	ExcludedDirectoryNames []GeneralSubtreeName
+	ExcludedEdiPartyNames  []GeneralSubtreeEdi
+	ExcludedRegisteredIDs  []GeneralSubtreeOid
 }
 
 type NameConstraintsJSON struct {
 	Critical bool `json:"critical"`
 
-	PermittedDNSDomains     []string    `json:"permitted_names,omitempty"`
-	PermittedEmailDomains   []string    `json:"permitted_email_addresses,omitempty"`
-	PermittedIPAddresses    []string    `json:"permitted_ip_addresses,omitempty"`
-	PermittedDirectoryNames []pkix.Name `json:"permitted_directory_names,omitempty"`
+	PermittedDNSDomains     []string            `json:"permitted_names,omitempty"`
+	PermittedEmailDomains   []string            `json:"permitted_email_addresses,omitempty"`
+	PermittedIPAddresses    []string            `json:"permitted_ip_addresses,omitempty"`
+	PermittedDirectoryNames []pkix.Name         `json:"permitted_directory_names,omitempty"`
+	PermittedEdiPartyNames  []pkix.EDIPartyName `json:"permitted_edi_party_names,omitempty"`
+	PermittedRegisteredIDs  []string            `json:"permitted_registred_id,omitempty"`
+
+	ExcludedDNSDomains     []string            `json:"excluded_names,omitempty"`
+	ExcludedEmailDomains   []string            `json:"excluded_email_addresses,omitempty"`
+	ExcludedIPAddresses    []string            `json:"excluded_ip_addresses,omitempty"`
+	ExcludedDirectoryNames []pkix.Name         `json:"excluded_directory_names,omitempty"`
+	ExcludedEdiPartyNames  []pkix.EDIPartyName `json:"excluded_edi_party_names,omitempty"`
+	ExcludedRegisteredIDs  []string            `json:"excluded_registred_id,omitempty"`
+}
 
-	ExcludedEmailDomains   []string    `json:"excluded_names,omitempty"`
-	ExcludedDNSDomains     []string    `json:"excluded_email_addresses,omitempty"`
-	ExcludedIPAddresses    []string    `json:"excluded_ip_addresses,omitempty"`
-	ExcludedDirectoryNames []pkix.Name `json:"excluded_directory_names,omitempty"`
+func (nc *NameConstraints) UnmarshalJSON(b []byte) error {
+	var ncJson NameConstraintsJSON
+	err := json.Unmarshal(b, &ncJson)
+	if err != nil {
+		return err
+	}
+	for _, dns := range ncJson.PermittedDNSDomains {
+		nc.PermittedDNSDomains = append(nc.PermittedDNSDomains, GeneralSubtreeString{Data: dns})
+	}
+	for _, email := range ncJson.PermittedEmailDomains {
+		nc.PermittedEmailDomains = append(nc.PermittedEmailDomains, GeneralSubtreeString{Data: email})
+	}
+	for _, ipNet := range ncJson.PermittedIPAddresses {
+		ip, ipnet, err := net.ParseCIDR(ipNet)
+		if err != nil {
+			return err
+		}
+		ipnet.IP = ip
```

Comment:
Ah, I see. We're slightly abusing the semantics of the object then, but I think that's OK because the semantics will only be abused when the certificate has an invalid CIDR on it, anyway. (192.168.1.1/24 isn't a valid CIDR).

tl;dr I'm OK with this.